### PR TITLE
Add Java data-plane sdk for easm service

### PR DESCRIPTION
**Pull request by thang-bit**
# Description

The SDK for Defender EASM Public Preview

[product documentation](https://learn.microsoft.com/en-us/azure/external-attack-surface-management/)
https://github.com/Azure/azure-sdk/issues/5062
Typespec PR: https://github.com/Azure/azure-rest-api-specs/pull/24937
arcboard: https://github.com/Azure/azure-sdk/issues/5062

Initial PR: https://github.com/Azure/azure-sdk-for-java/pull/36282

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/easm/azure-analytics-defender-easm/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+public final class ReadmeSamples {
+    public void readmeSamples() {
+        // BEGIN: com.azure.analytics.defender.easm.readme
+        String subscriptionId = Configuration.getGlobalConfiguration().get("SUBSCRIPTIONID");
+        String workspaceName = Configuration.getGlobalConfiguration().get("WORKSPACENAME");
+        String resourceGroupName = Configuration.getGlobalConfiguration().get("RESOURCEGROUPNAME");
+        String endpoint = Configuration.getGlobalConfiguration().get("ENDPOINT");
+
+        EasmClient easmClient = new EasmClientBuilder()
```
Comment (srnagar) [inline sdk/easm/azure-analytics-defender-easm/README.md:77]: Take a look at the README from App Configuration and add similar samples.
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/appconfiguration/azure-data-appconfiguration/README.md

File: sdk/easm/pom.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.azure</groupId>
+  <artifactId>azure-easm-service</artifactId>
+  <packaging>pom</packaging>
+  <version>1.0.0</version>
+  <modules>
+    <module>azure-analytics-defender-easm</module>
```
Comment (srnagar) [inline sdk/easm/pom.xml:12]: The [root pom](https://github.com/Azure/azure-sdk-for-java/blob/main/pom.xml) in the repo should also be updated as `easm` is a new directory.

**Comment by thang-bit** [comment]
Hi @srnagar, can you kindly review this PR whenever you have time? Please let me know if I should provide anything for review

**Review by srnagar — APPROVED** [review]
LGTM

**Comment by thang-bit** [comment]
Hi @srnagar, will this PR be merged automatically?

**Comment by haolingdong-msft** [comment]
Hi @thang-bit, the pipeline is incomplete because this is a new RP, and we should run the command to prepare pipeline. I have triggerd it for you.
The command can be found in the quickstart: https://github.com/Azure/azure-sdk-for-java/wiki/TypeSpec-Java-Quickstart#review-checklist
image

**Comment by haolingdong-msft** [comment]
Hi @thang-bit, please fix the CI issue.

File: sdk/easm/azure-analytics-defender-easm/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
+# Azure Easm client library for Java
+
+*Microsoft Defender External Attack Surface Management (Defender EASM)* continuously discovers and maps your digital attack surface to provide an external view of your online infrastructure. This visibility enables security and IT teams to identify unknowns, prioritize risk, eliminate threats, and extend vulnerability and exposure control beyond the firewall. Defender EASM leverages Microsoft’s crawling technology to discover assets that are related to your known online infrastructure, and actively scans these assets to discover new connections over time. Attack Surface Insights are generated by leveraging vulnerability and infrastructure data to showcase the key areas of concern for your organization.
+
+This package contains Microsoft Azure EASM Defender client library
```
Comment (haolingdong-msft) [inline sdk/easm/azure-analytics-defender-easm/README.md:5]:
```suggestion
This package contains Microsoft Azure EASM Defender client library.
```

File: sdk/easm/azure-analytics-defender-easm/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+</dependency>
+```
+[//]: # ({x-version-update-end})
+
+### Authentication
+
+[Azure Identity][azure_identity] package provides the default implementation for authenticating the client.
+
+## Key concepts
+### [Assets][assets_documentation]
```
Comment (haolingdong-msft) [inline sdk/easm/azure-analytics-defender-easm/README.md:38]: These are not linked to the correct links. Please update the links and make sure all the referenced links are valid.
image

File: sdk/easm/azure-analytics-defender-easm/assets.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "AssetsRepo" : "Azure/azure-sdk-assets",
+  "AssetsRepoPrefixPath" : "java",
+  "TagPrefix" : "java/easm/azure-analytics-defender-easm",
+  "Tag" : ""
```
Comment (haolingdong-msft) [inline sdk/easm/azure-analytics-defender-easm/assets.json:5]: Have you run the recording for the tests? The tag should not be empty.

**Comment by haolingdong-msft** [comment]
Hi @thang-bit, I don't see the readme get updated, but the comment is resolved.

**Comment by thang-bit** [comment]
@haolingdong-msft, thanks for reminding, I just pushed the new README

File: sdk/easm/azure-analytics-defender-easm/assets.json (hunk 5 above)
Comment (thang-bit) [reply sdk/easm/azure-analytics-defender-easm/assets.json:5]: @haolingdong-msft, I was able to run tests in both RECORD and PLAYBACK mode successfully. However, the assets.json file isn't updated. Is there any steps that I missed here?

File: .vscode/cspell.json [hunk 6]
```diff
@@ -480,6 +480,13 @@
         "Decibles"
       ]
     },
+    {
+      "filename": "sdk/easm/azure-analytics-defender-easm/**/*.md",
+      "words": [
+        "EASM",
+        "easm"
```
Comment (haolingdong-msft) [inline .vscode/cspell.json:487]: do we need both `EASM` and `easm`? in my memory, this is case insensitive.

**Comment by haolingdong-msft** [comment]
Hi @thang-bit, what is your expected time to release?

**Comment by thang-bit** [comment]
Hi @haolingdong-msft, I'm planning to complete the PR this sprint.

File: sdk/easm/azure-analytics-defender-easm/assets.json (hunk 5 above)
Comment (haolingdong-msft) [reply sdk/easm/azure-analytics-defender-easm/assets.json]: assets.json should be updated automatically by the testproxy when running in RECORD mode. Could you show us your  running command？

Comment (haolingdong-msft) [reply sdk/easm/azure-analytics-defender-easm/assets.json]: Add @samvaity  for testproxy related issue as well.

**Review by haolingdong-msft — APPROVED** [review]
LGTM, @weidongxu-microsoft to take a look as well.

**Review by weidongxu-microsoft — COMMENTED** [review]
There was azure-core version bump and pending azure-identity version bump.

Please wait for merge of https://github.com/Azure/azure-sdk-for-java/pull/37242

And then refresh main, call the python script to sync versions in pom.xml again.
https://github.com/Azure/azure-sdk-for-java/blob/main/eng/versioning/update_versions.py

File: sdk/easm/azure-analytics-defender-easm/README.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+[azure_subscription]: https://azure.microsoft.com/free/
+[azure_identity]: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity
+[discovery_runs_sample]: https://github.com/thang-bit/azure-sdk-for-java/blob/easm-java-sdk/sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/DiscoveryRunsSample.java
+[disco_template_sample]: https://github.com/thang-bit/azure-sdk-for-java/blob/easm-java-sdk/sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/DiscoTemplateSample.java
+[saved_filter_sample]: https://github.com/thang-bit/azure-sdk-for-java/blob/easm-java-sdk/sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/SavedFilterSample.java
+[external_ids_sample]: https://github.com/thang-bit/azure-sdk-for-java/blob/easm-java-sdk/sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/ManagingExternalIdsSample.java
+[logging]: https://github.com/Azure/azure-sdk-for-java/wiki/Logging-with-Azure-SDK
+[samples]: https://github.com/thang-bit/azure-sdk-for-java/tree/easm-java-sdk/samples
```
Comment (weidongxu-microsoft) [inline sdk/easm/azure-analytics-defender-easm/README.md:112]: you need a follow-up PR to update them to main branch

File: sdk/easm/azure-analytics-defender-easm/src/test/java/com/azure/analytics/defender/easm/generated/EasmClientTestBase.java
Comment (weidongxu-microsoft) [inline sdk/easm/azure-analytics-defender-easm/src/test/java/com/azure/analytics/defender/easm/generated/EasmClientTestBase.java:1]: if you use it, please move it out of generated package (up one folder level)

File: sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/DiscoveryRunsSample.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+        // In order to start discovery runs, we must first create a discovery group, which is a collection of known assets that we can pivot off of.
+        // These are created using the discoveryGroupsPut method
+
+        List<DiscoSource> seeds = Stream.concat(
+            Arrays.stream(Configuration.getGlobalConfiguration().get("HOSTS").split(","))
+                .map((host) -> new DiscoSource().setName(host).setKind(DiscoSourceKind.HOST)),
+            Arrays.stream(Configuration.getGlobalConfiguration().get("DOMAINS").split(","))
+                .map((domain) -> new DiscoSource().setName(domain).setKind(DiscoSourceKind.DOMAIN))
+        ).collect(Collectors.toList());
```
Comment (weidongxu-microsoft) [inline sdk/easm/azure-analytics-defender-easm/src/samples/java/com/azure/analytics/defender/easm/DiscoveryRunsSample.java:59]: Can we provide a sample of the List in sample code?

I am not sure whether customer would know what HOSTS and DOMAINS should be (they likely won't have them in their env variables).

**Comment by haolingdong-msft** [comment]
There supposed to have `tsp-location.yaml` file. It is usually automatically generated, but I don't see it in your pr. 

See document:
https://github.com/Azure/azure-sdk-for-java/wiki/TypeSpec-Java-Quickstart#configuration

You can refer to the [tsp-location.yaml](https://github.com/Azure/azure-sdk-tools/blob/main/doc/common/TypeSpec-Project-Scripts.md#tsp-locationyaml) which describes the supported properties in the file.

File: sdk/easm/azure-analytics-defender-easm/README.md (hunk 7 above)
Comment (thang-bit) [reply sdk/easm/azure-analytics-defender-easm/README.md:112]: Created a draft PR here: https://github.com/Azure/azure-sdk-for-java/pull/37256

**Comment by thang-bit** [comment]
Hi @haolingdong-msft, @weidongxu-microsoft, I added the typespec PR to the tsp-location.yaml file. It seems that it's complaining about the difference between the sdk generated from typespec file and the sdk in this PR.
Here are some of the modifications made in this PR:
- Added CountPagedResponse and CountPagedIterable since we have totalElements property in the page response
- Added cleanUp function to replace "+0000" in datetime to "+00:00" which is accepted in Java sdk

**Comment by weidongxu-microsoft** [comment]
@thang-bit 

I am updating the compare script to ignore files under `implementation` packages.
However, it is not recommended to modify code in `xxxClientImpl`, as when you add new feature, it had to be re-generated in future.
Seems only valid change there is the API related to `CountPagedResponse`, which may better to be done in a different class in `implementation` packages. We can take what you have, but we need to get a solution before your next api-version.

@haolingdong-msft for guide on partial-update for `xxxClient` class.

**Comment by thang-bit** [comment]
Thanks for your reviews @weidongxu-microsoft. 
- For the datetime format, we will definitely address that from the api in the next review. 
- For the CountPagedResponse, we added that to support the totalElements property. @azure-tools/typespec-java would just generate list operations with PagedResponse by default. I think, for the next java sdk version, we might be able to customize the typespec file for it to generate the CountPagedResponse instead

**Comment by thang-bit** [comment]
@weidongxu-microsoft, if there's no further comments, can we complete this PR?

**Comment by weidongxu-microsoft** [comment]
I've merged the PR to ignore the changes in implementation package.

But you still need to set partial-update and regen the SDK (you can then "git checkout" those under implementation package).
Please contact @haolingdong-msft if you need help.

**Comment by haolingdong-msft** [comment]
@thang-bit, since you modified the generated file, you need to enable `partial-update` support which will not overwrite your changes, and also as @weidongxu-microsoft  mentioned, [Verify TypeSpec Code Generation](https://dev.azure.com/azure-sdk/public/_build/results?buildId=3182173&view=logs&jobId=b70e5e73-bbb6-5567-0939-8415943fadb9&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=d31d5e2d-d8df-53bc-da94-87e19f5b2ee4) verifies SDK code is in sync with your TypeSpec source.

partial update guide: https://github.com/Azure/azure-sdk-for-java/wiki/TypeSpec-Java-Quickstart#partial-update

File: sdk/easm/azure-analytics-defender-easm/pom.xml [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <version>4.11.0</version> <!-- {x-version-update;org.mockito:mockito-core;external_dependency} -->
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-core-test</artifactId>
+      <version>1.22.0-beta.1</version> <!-- {x-version-update;com.azure:azure-core-test;dependency} -->
```
Comment (haolingdong-msft) [inline sdk/easm/azure-analytics-defender-easm/pom.xml:79]: why use beta version here? CI fails in the version validation: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3192085&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=bdeefc16-b669-5ebd-ad94-a2c19ade53b0

**Comment by thang-bit** [comment]
- @weidongxu-microsoft, @haolingdong-msft  all tests have passed. I'm waiting for this PR: https://github.com/Azure/azure-rest-api-specs/pull/26349 to be merged, then I can update the tsp-location.yml and cspell.json

**Comment by thang-bit** [comment]
@weidongxu-microsoft, I have a question, if we decide not to use the CountPagedResponse, but using the PagedResponse instead, how would users get total number of elements across pages?

**Comment by weidongxu-microsoft** [comment]
@thang-bit 

User would not able to get the number of total elements from `PagedResposne`.

Whether to include `CountPagedResponse` may depends
1. whether you expect customer to use it to decide code flow (e.g. when customer see 1000, do they stop paging?); or only to print out?
2. whether this be consistent across languages (does customer using Python or .NET can get it?)

The customization mentioned above is to allow you use the `CountPagedResponse` in `EasmClient` without it be overwritten by follow up codegen. So that if you decide to use it, you can.
Your previous change in `EasmClientImpl` with `CountPagedResponse` would need further consideration but if you decide to code it there, CI will no longer block you (as we excluded `implementation` package from that check).

File: sdk/easm/azure-analytics-defender-easm/tsp-location.yaml [hunk 10]
```diff
@@ -1,3 +1,3 @@
 directory: specification/riskiq/Easm
 commit: 219fcd35385531e2af9c5b210ea195b8d05b393b
```
Comment (weidongxu-microsoft) [inline sdk/easm/azure-analytics-defender-easm/tsp-location.yaml:2]: use the commit id on main branch history https://github.com/Azure/azure-rest-api-specs/commits/main
e.g. https://github.com/Azure/azure-rest-api-specs/commit/f5f584c65071dc0306544a195fbcdff156023040

**Review by weidongxu-microsoft — APPROVED** [review]
I am good with it for preview.

You can decide about `CountPagedResponse` vs `PagedResponse` anytime before GA.

Ping us if you want PR merged. (or do it yourself if you have permission)

**Comment by thang-bit** [comment]
@weidongxu-microsoft  I just take a look at our published python sdk, it seems that we used the PagedResponse pattern. So, I'm good with merging with this PR now. Thank you so much for your reviews

**Comment by thang-bit** [comment]
@weidongxu-microsoft, it seems that I don't have permission to merge this PR. Would you mind helping me merge it?

File: sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/models/CountPagedIterable.java
Comment (weidongxu-microsoft) [inline sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/models/CountPagedIterable.java:1]: @thang-bit I forget this. Remove the files that customer don't use. That include the `CountPaged*`, and maybe some others.

**Pull request merged**